observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -300,53 +317,85 @@ void logCancel(
 
-    private PayloadBuilder(T payload, int size) {
+    private PayloadBuilder(T payload, int size, boolean truncated) {
       this.payload = payload;
       this.size = size;
+      this.truncated = truncated;
     }
   }
 
   // TODO(dnvindhya): Create a unit test for the metadata conversion
-  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata) {
+  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata,
+      int maxHeaderBytes) {
     checkNotNull(metadata, "metadata");
+    checkArgument(maxHeaderBytes >= 0,
+        "maxHeaderBytes must be non negative");
     GrpcLogRecord.Metadata.Builder metadataBuilder = GrpcLogRecord.Metadata.newBuilder();
     // This code is tightly coupled with io.grpc.observabilitylog.v1.GrpcLogRecord.Metadata's
     // implementation
     byte[][] serialized = InternalMetadata.serialize(metadata);
+    boolean truncated = false;
     int totalMetadataBytes = 0;
     if (serialized != null) {
-      int singleMetadataEntryBytes = 0;
       // Calculate bytes for each GrpcLogRecord.Metadata.MetadataEntry
       for (int i = 0; i < serialized.length; i += 2) {
         String key = new String(serialized[i], Charsets.UTF_8);
         byte[] value = serialized[i + 1];
-        singleMetadataEntryBytes = totalMetadataBytes + key.length() + value.length;
+        if (NEVER_INCLUDED_METADATA.contains(key)) {
+          continue;
+        }
+        boolean forceInclude = ALWAYS_INCLUDED_METADATA.contains(key);
+        int metadataBytesAfterAdd = totalMetadataBytes + key.length() + value.length;
+        if (!forceInclude && metadataBytesAfterAdd > maxHeaderBytes) {
+          truncated = true;
+          continue;
+        }
         metadataBuilder.addEntryBuilder()
             .setKey(key)
             .setValue(ByteString.copyFrom(value));
-        totalMetadataBytes = singleMetadataEntryBytes;
+        if (!forceInclude) {
+          // force included keys do not count towards the size limit
+          totalMetadataBytes = metadataBytesAfterAdd;
+        }
       }
     }
-    return new PayloadBuilder<>(metadataBuilder, totalMetadataBytes);
+    return new PayloadBuilder<>(metadataBuilder, totalMetadataBytes, truncated);
   }
 
-  static PayloadBuilder<ByteString> createMesageProto(byte[] message) {
+  static PayloadBuilder<ByteString> createMesageProto(byte[] message, int maxMessageBytes) {
+    checkArgument(maxMessageBytes >= 0,
+        "maxMessageBytes must be non negative");
+    int desiredBytes = 0;
     int messageLength = message.length;
+    if (maxMessageBytes > 0) {
+      desiredBytes = Math.min(maxMessageBytes, messageLength);
+    }
     ByteString messageData =
-        ByteString.copyFrom(message, 0, messageLength);
-    return new PayloadBuilder<ByteString>(messageData, messageLength);
+        ByteString.copyFrom(message, 0, desiredBytes);
+
+    return new PayloadBuilder<ByteString>(messageData, messageLength,
```

Comment:
"Explicit type argument ByteString can be replaced with <> "

Reply:
Done.